libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,626 @@
+  using type = _T;
+};
+
+template <class _T, class... _Tp>
+struct __derived_from_pack {
+  constexpr static bool value =
+      __derived_from_pack<_T, typename __extract_last<_Tp...>::type>::value && __derived_from_pack<_Tp...>::value;
+};
+
+template <class _T, class _IterCategory>
+struct __derived_from_pack<_T, _IterCategory> {
+  constexpr static bool value = derived_from<_T, _IterCategory>;
+};
+
+template <class _View, class... _Views>
+struct __last_view : __last_view<_Views...> {};
+
+template <class _View>
+struct __last_view<_View> {
+  using type = _View;
+};
+
+template <class _Ref, class _RRef, class _It>
+concept __concat_indirectly_readable_impl = requires(const _It it) {
+  { *it } -> convertible_to<_Ref>;
+  { ranges::iter_move(it) } -> convertible_to<_RRef>;
+};
+
+template <class... _Rs>
+using __concat_reference_t = common_reference_t<range_reference_t<_Rs>...>;
+
+template <class... _Rs>
+using __concat_value_t = common_type_t<range_value_t<_Rs>...>;
+
+template <class... _Rs>
+using __concat_rvalue_reference_t = common_reference_t<range_rvalue_reference_t<_Rs>...>;
+
+template <class... _Rs>
+concept __concat_indirectly_readable =
+    common_reference_with<__concat_reference_t<_Rs...>&&, __concat_value_t<_Rs...>&> &&
+    common_reference_with<__concat_reference_t<_Rs...>&&, __concat_rvalue_reference_t<_Rs...>&&> &&
+    common_reference_with<__concat_rvalue_reference_t<_Rs...>&&, __concat_value_t<_Rs...> const&> &&
+    (__concat_indirectly_readable_impl<__concat_reference_t<_Rs...>,
+                                       __concat_rvalue_reference_t<_Rs...>,
+                                       iterator_t<_Rs>> &&
+     ...);
+
+template <class... _Rs>
+concept __concatable = requires {
+  typename __concat_reference_t<_Rs...>;
+  typename __concat_value_t<_Rs...>;
+  typename __concat_rvalue_reference_t<_Rs...>;
+} && __concat_indirectly_readable<_Rs...>;
+
+template <bool _Const, class... _Rs>
+concept __concat_is_random_access =
+    (random_access_range<__maybe_const<_Const, _Rs>> && ...) && (sized_range<__maybe_const<_Const, _Rs>> && ...);
+
+template <bool _Const, class... _Rs>
+concept __concat_is_bidirectional =
+    ((bidirectional_range<__maybe_const<_Const, _Rs>> && ...) && (common_range<__maybe_const<_Const, _Rs>> && ...));
+
+template <bool _Const, class... _Views>
+concept __all_forward = (forward_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... Ts>
+struct __apply_drop_first;
+
+template <bool _Const, class Head, class... Tail>
+struct __apply_drop_first<_Const, Head, Tail...> {
+  static constexpr bool value = (sized_range<__maybe_const<_Const, Tail>> && ...);
+};
+
+template <input_range... _Views>
+  requires(view<_Views> && ...) && (sizeof...(_Views) > 0) && __concatable<_Views...>
+class concat_view : public view_interface<concat_view<_Views...>> {
+  tuple<_Views...> views_;
+
+  template <bool _Const>
+  class iterator;
+  class sentinel;
```

Comment:
Exposition-only names are not standard, and thus should be __ugly (ditto below).

```suggestion
  template <bool _Const>
  class __iterator;
  class __sentinel;
```